File: Java/1053.java
__________________________________________________________________________________________________
sample 0 ms submission
class Solution {

    public int[] prevPermOpt1(int[] A) {
        int n = A.length, left = n - 2, right = n - 1, tmp;
        while (left >= 0  && A[left] <= A[left + 1]) left--;
        if (left < 0) return A;
        while (A[left] <= A[right]) right--;
        while (A[right - 1] == A[right]) right--;
        tmp = A[left]; A[left] = A[right]; A[right] = tmp;
        return A;
    }
}
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    public int[] prevPermOpt1(int[] A) {
        int n = A.length;
        TreeMap<Integer,Integer> map = new TreeMap<>();
        for(int i=n-1;i>=0;--i){
            if(!map.isEmpty()&&map.lowerKey(A[i])!=null){
                int key = map.lowerKey(A[i]);
                int id = map.get(key);
                int t = A[id];
                A[id] = A[i];
                A[i] = t;
                break;
            }
            else{
                map.put(A[i], i);
            }
        }
        return A;
    }
}
__________________________________________________________________________________________________

File: Python3/1053.py
__________________________________________________________________________________________________
sample 140 ms submission
class Solution:
    def prevPermOpt1(self, A: List[int]) -> List[int]:
        i = len(A) - 2
        while i >= 0:
            if A[i] <= A[i + 1]:
                i -= 1
            else:
                break
        if i < 0:
            return A
        j = len(A) - 1
        while A[j] >= A[i]:
            j -= 1
        while A[j - 1] == A[j]:
            j -= 1
        A[i], A[j] = A[j], A[i]
        return A        
__________________________________________________________________________________________________
sample 144 ms submission
class Solution:
    def prevPermOpt1(self, A: List[int]) -> List[int]:
        n = len(A)
        for left in range(n-2, -1, -1):
            if A[left] > A[left+1]:
                break
        else:
            return A
        right = A.index(max(a for a in A[left+1:] if a < A[left]), left)
        A[left], A[right] = A[right], A[left]
        return A
__________________________________________________________________________________________________